Feature/4 get more than 300 candles

Cattes opened the pull request:
#4 Implement History class to be able to get more than 300 candles from the past.

teleprint-me commented:
obviously the build is failing... I'm not going to merge a broken build into a stable branch.

I only skimmed the code because I'm busy with other things ATM. the build log details that the call datetime.fromisoformat is invalid. 

not sure if you meant to call the start object by using start.fromisoformat which should be a datetime object.

also, make sure you're confident in your implementation before pushing future PRs; meaning you personally tested it, are able to execute and utilize it, and that the tests pass (even if the test is crude).

the errors seem to be related mostly to the tests though. I'd feel more confident in accepting the PR if the tests were passing and knowing you personally battle tested the methods; meaning it works on your machine at minimum.

the rest can be figured out as we go along.

Cattes commented:
Hi, sry I did not check the CICD pipeline after creating the PR.

It seems the `datetime.datetime.fromisoformat` [was only implemented in python 3.7](https://docs.python.org/3/library/datetime.html#datetime.date.fromisoformat), which is why the build is failling for python3.6. I only tested it locally with python3.8 [where the CICD pipeline is also succeeding here](https://travis-ci.com/github/teleprint-me/coinbasepro-python/jobs/519255459).

I will have a look at it and modify the code so that it also works for python3.6

Cattes commented:
The easiest solution would be to expect the `start` and `end` values as `datetime.datetime` objects and not isoformatted strings which I transform to datetime. But this would break from the example in the readme that all [work with isoformatted strings](https://github.com/teleprint-me/coinbasepro-python#cbpropublicproductshistory)(which is matching the input the cbpro api expects)

teleprint-me commented:
iso8601 library can parse the iso format result and return a datetime string. this can fix the issue of breaking changes and would allow you to send the expected api input.

Cattes commented:
> [...] the con to this is it would create another dependency.

Yes that is why I did not add `arrow` or another library first, but added a custom function for parsing.
But the custom function has more weaknesses than just using an external library. 

So if you are ok with an external library we can keep the current solution.

Otherwise we could keep my `dt_string_to_datetime` function from this commit https://github.com/teleprint-me/coinbasepro-python/pull/6/commits/7198e9ef2f3ee05c33d6402a9491cee91c3c870d

Or move away from the current way of expecting a string from the uer and for the History class the user would have to supply a datetime.datetime object directly.

You can decide.

teleprint-me commented:
I'm okay with adding iso8601 as a dep. it leaves the interface intact and doesn't require any changes to what the methods expect which avoids a lot of the pitfalls of doing otherwise.

just make sure to update the minor and revision version and add your handle or name to the contributors.txt if you want credit for your contribution. that way I can merge your addition into the master branch. 

I'll add a tag and we can watch to see how things play out for others interested in using History class. I can also add it to the docs at some point. 

I was thinking about creating a docs directory and adding markdown files for each topic in that directory instead of just having one massive readme.

I'm just focused on my web app at the moment and that's taking higher priority right now. I'm only able to work on it a few days a week, so most of my free time is being spent on that.

thanks also for taking the time to put all this together. any future contributions are always more than welcome. lmk if you're interested in becoming a maintainer. I could use the help. I wasn't planning on taking over for this project and am in kind of an awkward situation 🙃. 

I just happened to see something that could be improved and wanted to contribute back to the original project. things haven't really played out the way I expected. I'm still hoping to hear back from the original dev to see what they want to do.

if I do take over at some point, then I'm going to need help to keep up with it. that's just the reality of the situation.

Cattes commented:
> I'm okay with adding iso8601 as a dep. it leaves the interface intact and doesn't require any changes to what the methods expect which avoids a lot of the pitfalls of doing otherwise.

Alright. I agree, and it is much more fault tolerant than my crude function.


> just make sure to update the minor and revision version and add your handle or name to the contributors.txt if you want credit for your contribution. that way I can merge your addition into the master branch.

I updated the version from 2.1.28 to 2.2.0 and updated the contributors.txt

> I'll add a tag and we can watch to see how things play out for others interested in using History class. I can also add it to the docs at some point.
> 
> I was thinking about creating a docs directory and adding markdown files for each topic in that directory instead of just having one massive readme.

I added a section to the long readme for the History class with an example. 

> I'm just focused on my web app at the moment and that's taking higher priority right now. I'm only able to work on it a few days a week, so most of my free time is being spent on that.
> 
> thanks also for taking the time to put all this together. any future contributions are always more than welcome. lmk if you're interested in becoming a maintainer. I could use the help. I wasn't planning on taking over for this project and am in kind of an awkward situation 🙃.
> 
> I just happened to see something that could be improved and wanted to contribute back to the original project. things haven't really played out the way I expected. I'm still hoping to hear back from the original dev to see what they want to do.
> 
> if I do take over at some point, then I'm going to need help to keep up with it. that's just the reality of the situation.

Yes I also hope the original project owner will resurface and take over again.
I am not yet familiar with working with websockts and hope to gather experience with them using this project. If I find other Issues or feature requests I will open a ticket for them 🙂 

As for helping to maintain the package, I have limited time too but I would be happy to gather some more experience contributing to the open source community.
